src/codegen/optimizer.lisp

```diff
@@ -292,6 +299,113 @@
+           (type tc:environment env)
+           (values node &optional))
+  (labels ((inline-application (node &rest rest)
+             (declare (ignore rest)
+                      (dynamic-extent rest))
+             (let ((rator (node-application-rator node))
+                   (rands (node-application-rands node)))
+               (when (node-variable-p rator)
+                 (let* ((name (node-variable-value rator))
+                        (code (tc:lookup-code env name :no-error t))
+                        (vars (and (node-abstraction-p code)
+                                   (node-abstraction-vars code)))
+                        (inline-p (and (node-abstraction-p code)
+                                       (tc:function-env-entry-inline-p
+                                        (tc:lookup-function env name)))))
+                   (cond ((null rands)
+                          (make-node-variable
+                           :type (node-type node)
+                           :value name))
+                         ((and code
+                               inline-p
+                               ;; TODO: Optimize the case when the lengths are unequal
+                               (= (length vars)
+                                  (length rands)))
+                          (multiple-value-bind (bindings subs)
+                              (loop :for var :in vars
+                                    :for val :in rands
+                                    :for new-var := (gentemp (symbol-name var))
+                                    :collect (cons new-var val) :into bindings
+                                    :collect (make-ast-substitution
+                                              :from var
+                                              :to (make-node-variable
+                                                   :type (node-type val)
+                                                   :value new-var))
+                                      :into subs
+                                    :finally (return (values bindings subs)))
+                            (make-node-let
+                             :type (node-type node)
+                             :bindings bindings
+                             :subexpr (apply-ast-substitution
+                                       subs
+                                       (node-abstraction-subexpr code)))))
+                         (t
+                          (make-node-application
+                           :type (node-type node)
+                           :rator (make-node-variable
+                                   :type (tc:make-function-type*
+                                          (mapcar #'node-type rands)
+                                          (node-type node))
+                                   :value name)
+                           :rands rands)))))))
+           ;; FIXME: What kind of frontend code evokes DIRECT-APPLICATION
```

Comment:
[This pass](https://github.com/coalton-lang/coalton/blob/main/src/codegen/optimizer.lisp#L107) turns regular calls into direct calls when it's known that regular function calls can be used.